File: r/day5_part2_2024.R

# Function to check if an update is in the correct order
is_correct_order <- function(update, rules) {
  for (i in 1:(length(update) - 1)) {
    for (j in (i + 1):length(update)) {
      rule_key <- paste(update[j], update[i], sep = "|")
      if (rule_key %in% names(rules)) {
        return(FALSE) # Found a violation
      }
    }
  }
  return(TRUE)
}

# Function to topologically sort an update based on rules (Kahn's Algorithm)
topo_sort <- function(update, rules) {
    
  in_degree <- integer(0)
  
  #Calculate in-degrees considering only nodes present in the current update
  for (page in update){
      in_degree[as.character(page)] <- 0
  }
  for (rule in names(rules)) {
      
    parts <- strsplit(rule, "\\|")[[1]]
    
    if(parts[1] %in% update & parts[2] %in% update){
        in_degree[parts[2]] <-  in_degree[parts[2]] + 1
    }
  }

  queue <- names(in_degree)[in_degree == 0]
  sorted_update <- character(0)

  while (length(queue) > 0) {
    u <- queue[1]
    queue <- queue[-1]
    sorted_update <- c(sorted_update, u)

      for (v in update) {
          rule_key <- paste(u,v,sep="|")
          if(rule_key %in% names(rules)){
              
              if(in_degree[as.character(v)] > 0){
                in_degree[as.character(v)] <- in_degree[as.character(v)] -1
                if (in_degree[as.character(v)] == 0) {
                  queue <- c(queue, as.character(v))
                }
              }
          }
      }
  }
    
  if(length(sorted_update) != length(update)){
      return(NULL) #Cycle detected, should not occur given problem constraints, but handled for completeness
  }

  return(sorted_update)
}


# Main function
main <- function() {
  # Read input from file
  input <- readLines("input.txt")
  
  # Separate rules and updates
  rules_end_index <- which(input == "") -1
  rules_lines <- input[1:rules_end_index]
  updates_lines <- input[(rules_end_index + 2):length(input)]

  # Create a rules dictionary (hash map)
  rules <- list()
  for (rule in rules_lines) {
    rules[[rule]] <- TRUE
  }

  # Process updates
  correct_updates_middle_sum <- 0
  incorrect_updates_middle_sum <- 0

  for (update_line in updates_lines) {
    update <- strsplit(update_line, ",")[[1]]
    
    if (is_correct_order(update, rules)) {
      middle_index <- (length(update) + 1) %/% 2
      correct_updates_middle_sum <- correct_updates_middle_sum + as.integer(update[middle_index])
    } else{
        sorted_update <- topo_sort(update, rules)
        if(!is.null(sorted_update)){
          middle_index <- (length(sorted_update) + 1) %/% 2
          incorrect_updates_middle_sum <- incorrect_updates_middle_sum + as.integer(sorted_update[middle_index])
        }
    }
  }

  # Print results
  cat("Sum of middle page numbers of correctly-ordered updates:", correct_updates_middle_sum, "\n")
  cat("Sum of middle page numbers of incorrectly-ordered updates (after sorting):", incorrect_updates_middle_sum, "\n")
}

# Run the main function
main()

File: matlab/day5_part2_2024.m

function main()
    % Reads input from input.txt and prints the output to standard output.
    % Solves the "Print Queue" coding challenge.

    % Read the entire file content
    fileContent = fileread('input.txt');

    % Split the content into rules and updates
    parts = strsplit(fileContent, '\n\n');
    rulesStr = strsplit(parts{1}, '\n');
    updatesStr = strsplit(parts{2}, '\n');

    % Parse the rules into a dependency graph (adjacency list)
    % and also store the reverse dependencies for easier checking.
    dependencies = containers.Map('KeyType', 'char', 'ValueType', 'any');
    reverseDependencies = containers.Map('KeyType', 'char', 'ValueType', 'any');
    allPages = containers.Map('KeyType', 'char', 'ValueType', 'logical');

    for i = 1:length(rulesStr)
        if isempty(rulesStr{i})
            continue;
        end
        ruleParts = strsplit(rulesStr{i}, '|');
        pre = ruleParts{1};
        post = ruleParts{2};

        % Add to allPages to keep track of all unique page numbers mentioned in rules
        allPages(pre) = true;
        allPages(post) = true;

        % Store dependencies: pre must come before post
        if ~isKey(dependencies, pre)
            dependencies(pre) = {};
        end
        dependencies(pre) = [dependencies(pre), {post}];

        % Store reverse dependencies: post must come after pre
        if ~isKey(reverseDependencies, post)
            reverseDependencies(post) = {};
        end
        reverseDependencies(post) = [reverseDependencies(post), {pre}];
    end

    % Process each update
    correctlyOrderedSum = 0;
    incorrectlyOrderedSum = 0;

    for i = 1:length(updatesStr)
        if isempty(updatesStr{i})
            continue;
        end
        currentUpdatePages = strsplit(updatesStr{i}, ',');
        currentUpdatePages = strtrim(currentUpdatePages); % Trim whitespace

        % Filter rules to only consider pages present in the current update
        relevantDependencies = containers.Map('KeyType', 'char', 'ValueType', 'any');
        relevantReverseDependencies = containers.Map('KeyType', 'char', 'ValueType', 'any');
        updatePageSet = containers.Map('KeyType', 'char', 'ValueType', 'logical');
        for j = 1:length(currentUpdatePages)
            updatePageSet(currentUpdatePages{j}) = true;
        end

        for page = keys(dependencies)
            pageStr = page{1};
            if isKey(updatePageSet, pageStr)
                dependentPages = dependencies(pageStr);
                for k = 1:length(dependentPages)
                    dependentPage = dependentPages{k};
                    if isKey(updatePageSet, dependentPage)
                        if ~isKey(relevantDependencies, pageStr)
                            relevantDependencies(pageStr) = {};
                        end
                        relevantDependencies(pageStr) = [relevantDependencies(pageStr), {dependentPage}];
                    end
                end
            end
        end

        for page = keys(reverseDependencies)
            pageStr = page{1};
            if isKey(updatePageSet, pageStr)
                precedentPages = reverseDependencies(pageStr);
                for k = 1:length(precedentPages)
                    precedentPage = precedentPages{k};
                    if isKey(updatePageSet, precedentPage)
                        if ~isKey(relevantReverseDependencies, pageStr)
                            relevantReverseDependencies(pageStr) = {};
                        end
                        relevantReverseDependencies(pageStr) = [relevantReverseDependencies(pageStr), {precedentPage}];
                    end
                end
            end
        end

        % Check if the current update is in the correct order
        isCorrect = true;
        for j = 1:length(currentUpdatePages)
            currentPage = currentUpdatePages{j};
            % Check if any page that should come *after* currentPage is before it
            if isKey(relevantDependencies, currentPage)
                pagesAfter = relevantDependencies(currentPage);
                for k = 1:length(pagesAfter)
                    pageAfter = pagesAfter{k};
                    % Find the index of pageAfter in the current update
                    idxPageAfter = find(strcmp(currentUpdatePages, pageAfter), 1);
                    if ~isempty(idxPageAfter) && idxPageAfter < j
                        isCorrect = false;
                        break;
                    end
                end
            end
            if ~isCorrect, break; end

            % Check if any page that should come *before* currentPage is after it
            if isKey(relevantReverseDependencies, currentPage)
                pagesBefore = relevantReverseDependencies(currentPage);
                for k = 1:length(pagesBefore)
                    pageBefore = pagesBefore{k};
                    % Find the index of pageBefore in the current update
                    idxPageBefore = find(strcmp(currentUpdatePages, pageBefore), 1);
                    if ~isempty(idxPageBefore) && idxPageBefore > j
                        isCorrect = false;
                        break;
                    end
                end
            end
            if ~isCorrect, break; end
        end

        % Calculate middle page number
        numPages = length(currentUpdatePages);
        middleIndex = ceil(numPages / 2);
        middlePage = str2double(currentUpdatePages{middleIndex});

        if isCorrect
            correctlyOrderedSum = correctlyOrderedSum + middlePage;
        else
            % If not correct, sort it using topological sort (Kahn's algorithm)
            sortedUpdate = topologicalSort(currentUpdatePages, relevantDependencies, relevantReverseDependencies);
            numSortedPages = length(sortedUpdate);
            middleIndexSorted = ceil(numSortedPages / 2);
            middlePageSorted = str2double(sortedUpdate{middleIndexSorted});
            incorrectlyOrderedSum = incorrectlyOrderedSum + middlePageSorted;
        end
    end

    fprintf('Sum of middle pages of correctly ordered updates: %d\n', correctlyOrderedSum);
    fprintf('Sum of middle pages of correctly ordered incorrectly ordered updates: %d\n', incorrectlyOrderedSum);
end

function sortedPages = topologicalSort(pages, dependencies, reverseDependencies)
    % Performs a topological sort on a list of pages given dependency rules.
    % Uses Kahn's algorithm.

    % Initialize in-degrees for all pages in the current update
    inDegree = containers.Map('KeyType', 'char', 'ValueType', 'double');
    for i = 1:length(pages)
        inDegree(pages{i}) = 0;
    end

    % Calculate in-degrees based on relevant dependencies
    for page = keys(reverseDependencies)
        pageStr = page{1};
        if isKey(inDegree, pageStr) % Ensure the page is in the current update
            precedents = reverseDependencies(pageStr);
            for k = 1:length(precedents)
                precedentPage = precedents{k};
                if isKey(inDegree, precedentPage) % Ensure the precedent is in the current update
                    inDegree(pageStr) = inDegree(pageStr) + 1;
                end
            end
        end
    end

    % Initialize a queue with all nodes having an in-degree of 0
    queue = {};
    for i = 1:length(pages)
        if inDegree(pages{i}) == 0
            queue{end+1} = pages{i};
        end
    end

    sortedPages = {};

    % Process the queue
    while ~isempty(queue)
        % Dequeue a node
        currentPage = queue{1};
        queue(1) = [];

        % Add to the sorted list
        sortedPages{end+1} = currentPage;

        % Decrease in-degree of adjacent nodes
        if isKey(dependencies, currentPage)
            dependents = dependencies(currentPage);
            for k = 1:length(dependents)
                dependentPage = dependents{k};
                if isKey(inDegree, dependentPage) % Ensure the dependent is in the current update
                    inDegree(dependentPage) = inDegree(dependentPage) - 1;
                    % If in-degree becomes 0, enqueue it
                    if inDegree(dependentPage) == 0
                        queue{end+1} = dependentPage;
                    end
                end
            end
        end
    end

    % Check for cycles (though the problem implies no cycles in valid inputs)
    if length(sortedPages) ~= length(pages)
        error('Graph has a cycle, cannot perform topological sort.');
    end
end

% To run this program:
% 1. Save the code as a .m file (e.g., print_queue.m).
% 2. Create a file named input.txt in the same directory with your puzzle input.
% 3. Run the script from the MATLAB command window by typing: main
